jenkins_jobs/urllib_kerb.py

```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env python
+# Copyright (C) 2015 OpenStack Foundation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import re
+import logging
+from six.moves.urllib.error import HTTPError
+from six.moves.urllib.request import BaseHandler, build_opener
+import kerberos
+
+logger = logging.getLogger(__name__)
+
+
+class HTTPNegotiateHandler(BaseHandler):
+    handler_order = 490  # before Digest auth
+
+    def __init__(self, max_tries=5):
+        self.krb_context = None
+        self.tries = 0
+        self.max_tries = max_tries
+        self.re_extract_auth = re.compile('.*?Negotiate\s*([^,]*)', re.I)
+
+    def http_error_401(self, req, _, __, ___, headers):
+        logger.debug("INSIDE http_error_401")
+        try:
+            try:
+                krb_req = self._extract_krb_value(headers)
+            except ValueError:
+                # Negotiate header not found or a similiar error
+                # we can't handle this, let the next handler have a go
+                return None
+
+            if not krb_req:
+                # First reply from server (no neg value)
+                self.tries = 0
+                krb_req = ""
```

Comment:
I prefer to be explicit here.

In case self._extract_krb_value returned None, it is guaranteed to be an empty string now (which can be passed to kerberos.authGSSClientStep).